R/xgb_binning.R

```diff
@@ -0,0 +1,250 @@
+#'   new step added to the sequence of existing steps (if any).
+#' @keywords binning
+#' @concept preprocessing
+#' @concept xgboost
+#' @concept discretization
+#' @concept factors
+#' @export
+#' @details step_xgb_binning creates non-uniform bins from numerical
+#'  variables by utilizing the information about the outcome
+#'  variable and applying the xgboost model. It is advised to impute
+#'  missing values before this step. This step is intented to be
+#'  used particularly with linear models because thanks to creating
+#'  non-uniform bins it becomes easier to learn non-linear patterns
+#'  from the data.
+#'
+#' This step requires the \pkg{xgboost} package. If not installed, the
+#'  step will stop with a note about installing the package.
+#'
+#' Note that the original variables will be replaced with the new bins.
+#'  The new variables will have names that begin with `prefix` followed
+#'  by their original name.
+#'
+#' @examples
+#' data(credit_data)
+#' library(rsample)
+#'
+#' split <- initial_split(credit_data, strata = "Status")
+#'
+#' credit_data_tr <- training(split)
+#' credit_data_te <- testing(split)
+#'
+#' xgb_rec <- recipe(Status ~ ., data = credit_data_tr) %>%
+#'   step_medianimpute(all_numeric()) %>%
+#'   step_xgb_binning(all_numeric(), outcome = "Status")
+#'
+#' xgb_rec <- prep(xgb_rec, training = credit_data_tr, retain = TRUE)
+#'
+#' xgb_test_bins <- bake(xgb_rec, credit_data_te)
+#' @seealso [recipe()] [prep.recipe()] [bake.recipe()]
+#'
+
+step_xgb_binning <-
+  function(recipe,
+           ...,
+           role = NA,
+           trained = FALSE,
+           outcome = NULL,
+           rules = NULL,
+           prefix = "xgb_bin",
+           skip = FALSE,
+           id = rand_id("xgb_binning")) {
+    if (is.null(outcome))
+      stop("`outcome` should select at least one column.", call. = FALSE)
+
+    # recipes_pkg_check("xgb_binninb") - TO DO CHECKS
+
+    add_step(
+      recipe,
+      step_xgb_binning_new(
+        terms = ellipse_check(...),
+        role = role,
+        trained = trained,
+        outcome = outcome,
+        rules = rules,
+        prefix = prefix,
+        skip = skip,
+        id = id
+      )
+    )
+  }
+
+step_xgb_binning_new <-
+  function(terms, role, trained, outcome, rules,
+           prefix, skip, id) {
+    step(
+      subclass = "xgb_binning",
+      terms = terms,
+      role = role,
+      trained = trained,
+      outcome = outcome,
+      rules = rules,
+      prefix = prefix,
+      skip = skip,
+      id = id
+    )
+  }
+
+run_xgboost <- function(.train, .test, .max_bin){
+  xgboost::xgb.train(
+    params = list(
+      eta = 0.3,
+      max_bin = .max_bin,
+      max_depth = 1,
+      min_child_weight = 5
+    ),
+    nrounds = 100,
+    data = .train,
+    watchlist = list(
+      train = .train,
+      test  = .test
+    ),
+    tree_method = "hist",
+    early_stopping_rounds = 10,
+    objective = "binary:logistic",
```

Comment:
This could conceivably be used for regression models too (right?) so maybe set this based on the class of the original outcome vector.

Reply:
That's definitely a good point and I would propose something like this since xgboost cares only about the number of levels of the outcome and actually doesn't run on factor outcomes:

```
# Checking the number of levels of the outcome
  levels <- sort(unique(df[[outcome]]))

  if(any(is.na((levels)))){
    stop("Outcome variable contains missing values.", call. = FALSE)
  }

  # Defining the objective function
  objective <- if (length(levels) == 2){
    "binary:logistic"
  } else if (class(df[[outcome]]) %in% c("double", "integer")){
    "reg:squarederror"
  } else {
    stop("Outcome variable needs to have two levels (binary classification) 
         or be a double/ integer (regression)", call. = FALSE)
  }
```